src/Tree/Tree.js

```diff
@@ -1,238 +1,330 @@
+                className={className}
+                role={level === 0 ? 'tree' : 'group'}>
+                {
+                    React.Children.map(children, (child) => {
+                        return React.cloneElement(child, {
+                            expandData,
+                            level,
+                            onExpandClick
+                        });
+                    })
                 }
-                return (
-                    <div className='fd-tree__col' key={index}>
-                        {typeof element === 'object' ? (
-                            <a className='fd-has-font-weight-semi' href={element.linkUrl}>
-                                {element.displayText ? element.displayText : element.linkUrl}
-                            </a>
-                        ) : (
-                            element
-                        )}
-                    </div>
-                );
-            });
+            </ul>
+        );
+    }
+}
 
-            let tree;
-
-            let displayLevel =
-            'fd-tree__group fd-tree__group--sublevel-' + depthLevel;
-
-            if (row.hasChildren && this.state.iStates[row.id]) {
-                tree = this.createTreeList(row.children, true, depthLevel + 1);
-            }
-            if (isChild) {
-                return (
-                    <ul className={displayLevel} key={row.id}
-                        role='group'>
-                        <ul className='fd-tree-child'>
-                            <li
-                                aria-expanded='true'
-                                className='fd-tree__item'
-                                key={row.id}
-                                role='treeitem'>
-                                <div className='fd-tree__row'>
-                                    {parent}
-                                    {<Dropdown />}
-                                </div>
-                                {tree}
-                            </li>
-                        </ul>
-                    </ul>
-                );
-            }
-            depthLevel = 0;
-            return (
-                <li
-                    aria-expanded='true'
-                    className='fd-tree__item'
-                    key={row.id}
-                    role='treeitem'>
-                    <div className='fd-tree__row'>
-                        {parent}
-                        {<Dropdown />}
-                    </div>
-                    {tree}
-                </li>
-            );
+TreeList.propTypes = {
+    children: PropTypes.node,
+    expandData: PropTypes.object,
+    isExpanded: PropTypes.bool,
+    level: PropTypes.number,
+    onExpandClick: PropTypes.func
+};
+
+TreeList.defaultProps = {
+    level: 0
+};
+
+export class Tree extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            expandData: {},
+            isExpandAll: false
```

Comment:
I would caution against tracking too much in state (if any) in this component.  It is very easy to either get out of sync it prevents the consumer from making any alterations with props.  Because this is a component library, in large part, these components should be "dumb".  Their job is simply to render what the consumer tells them to and have some callbacks exposed that tell the consumer if anything changed.  They should look for the consumer to maintain any sort of "state" because they likely have a redux store to leverage.

Reply:
I agree, but I feel like the expand/collapse state is something that the consumer shouldn't have to manage through application state. It does feel like something that the component should handle and the application shouldn't have to worry about.

Reply:
I don't disagree, however a consumer _can't_ control it now even if they wanted to.  If we're going to track this type of state, then we need to expose a prop and add a `getDerivedStateFromProps` (formerly `componentWillReceiveProps`) function to sync state with incoming props.  This way, a consumer _could_ provide an initial or updated state.  This would probably also mean exposing a callback prop for when the expansion state changes.

I was thinking that maybe it would be easier to just have `TreeHeader` and `TreeList` maintain their own expanded states, but the expand/collapse-all may not play nice with that so we can leave it where it is.

I still think this type of state shouldn't be in this component, but the `Tree` component is still under review and I've heard there may actually be multiple tree-type components coming so there will be time to revisit this again.

This is certainly _MUCH_ better than it was before! 👍

Reply:
Also, if we expose props to let consumers manage expand states, then we likely need to expose `id` props for the items.  The component could use `shortid` to generate ones if none are provided, but consumers will need to provide these if they want to work with them in their own state management.

Reply:
Allowed the consumer to override the expand/collapse state of both the `isExpandAll` variable and `isExpanded` for `TreeItem`s in 62ccf9c.
`isExpandAll` Ex: 
```
<TreeView isExpandAll>
    <TreeHead>
        <TreeCol>Header 1</TreeCol>
    </TreeHead>
    <Tree>
        .....
    </Tree>
</TreeView>
```

`isExpanded` Ex:
```
<TreeView isExpandAll>
    <TreeHead>
        ...
    </TreeHead>
    <Tree>
        <TreeItem isExpanded>
            <TreeRow>
                <TreeCol />
            </TreeRow>
            <TreeBranch>
                <TreeItem>
                    <TreeRow>
                         <TreeCol />
                    </TreeRow>
                </TreeItem>
            </TreeBranch>
        </TreeItem>
    </Tree>
</TreeView>
```